maven-embedder/src/main/java/org/apache/maven/cli/props/InterpolationHelper.java

```diff
@@ -0,0 +1,364 @@
+     * specified value. If the specified value contains the syntax
+     * {@code ${&lt;prop-name&gt;}}, where {@code &lt;prop-name&gt;}
+     * refers to either a configuration property or a system property,
+     * then the corresponding property value is substituted for the variable
+     * placeholder. Multiple variable placeholders may exist in the
+     * specified value as well as nested variable placeholders, which
+     * are substituted from inner most to outer most. Configuration
+     * properties override system properties.
+     * </p>
+     *
+     * @param val The string on which to perform property substitution.
+     * @param currentKey The key of the property being evaluated used to
+     *        detect cycles.
+     * @param cycleMap Map of variable references used to detect nested cycles.
+     * @param configProps Set of configuration properties.
+     * @return The value of the specified string after system property substitution.
+     * @throws IllegalArgumentException If there was a syntax error in the
+     *         property placeholder syntax or a recursive variable reference.
+     **/
+    public static String substVars(
+            String val, String currentKey, Map<String, String> cycleMap, Map<String, String> configProps)
+            throws IllegalArgumentException {
+        return substVars(val, currentKey, cycleMap, configProps, (SubstitutionCallback) null);
+    }
+
+    /**
+     * <p>
+     * This method performs property variable substitution on the
+     * specified value. If the specified value contains the syntax
+     * {@code ${&lt;prop-name&gt;}}, where {@code &lt;prop-name&gt;}
+     * refers to either a configuration property or a system property,
+     * then the corresponding property value is substituted for the variable
+     * placeholder. Multiple variable placeholders may exist in the
+     * specified value as well as nested variable placeholders, which
+     * are substituted from inner most to outer most. Configuration
+     * properties override system properties.
+     * </p>
+     *
+     * @param val The string on which to perform property substitution.
+     * @param currentKey The key of the property being evaluated used to
+     *        detect cycles.
+     * @param cycleMap Map of variable references used to detect nested cycles.
+     * @param configProps Set of configuration properties.
+     * @param callback the callback to obtain substitution values
+     * @return The value of the specified string after system property substitution.
+     * @throws IllegalArgumentException If there was a syntax error in the
+     *         property placeholder syntax or a recursive variable reference.
+     **/
+    public static String substVars(
+            String val,
+            String currentKey,
+            Map<String, String> cycleMap,
+            Map<String, String> configProps,
+            SubstitutionCallback callback)
+            throws IllegalArgumentException {
```

Comment:
The throws in the signature is redundant for runtime exceptions and we have good Javadoc here.

Reply:
I'll fix that.  Fwiw, those three classes come from Karaf/Felix.  They even provide a bit more, as they can even be used to programmatically modify properties file while keeping layout and comments (though I don't think we'll need that in the short term), or even typed properties (not included here).

Reply:
Fixed